jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java

```diff
@@ -125,7 +130,37 @@
    */
   @Nullable
   T call() throws IOException, RegistryException {
-    return call(initialRequestUrl);
+    return callWithInsecureRegistryHandling(initialRequestUrl);
```

Comment:
Hmm, I feel like the checking should be done here for whether or not to handle insecure registries based on the flag so that it launches two codepaths - one for calling without allowing insecure registries, and one for calling with allowing insecure registries - rather than have that logic be encapsulated in `callWithInsecureRegistryHandling`. From the function call here, it seems like insecure registry handling is the default.

Reply:
> it launches two codepaths

It's not exactly like there will be two codepaths based on `allowInsecureRegistries`. They cross-over, for example, you can first contact an HTTP server (failover-to-HTTP occurred) which redirects the client to a secure HTTPS server. With `callWithInsecureRegistryHandling()`, you don't have to think about those things anymore. You can pass a URL to `callWithInsecureRegistryHandling()` recursively and it will handle things correctly.

> From the function call here, it seems like insecure registry handling is the default.

You are saying the naming `callWithInsecureRegistryHandling` is giving you the impression that it will attempt insecure communication, right? I'll think about a better name, but if you have some suggestions, let me know.

Reply:
Ah yes, I'm thinking about this more in the lines of having `allowInsecureRegistries` be checked in one location to launch separate codepaths, one for secure and one for insecure. The one for insecure will be essentially the one for secure, except with additional functionality for handling HTTPS connect/TLS verification errors (codepaths with cross-over, but a different entrypoint). So, rather than checking `allowInsecureRegistries` in multiple places in `callWithInsecureRegistryHandling`, a single check would determine if the insecure handling code entrypoint would be launched vs the secure handling code entrypoint. I think I need to think more about this too.

Reply:
> So, rather than checking allowInsecureRegistries in multiple places in callWithInsecureRegistryHandling, a single check would determine if the insecure handling code entrypoint would be launched vs the secure handling code entrypoint.

I think I got what you mean. That makes sense too. I'll try your idea and push up changes, and then we can evaluate which one will look better.

Reply:
Here's the attempt with your idea, if I understood it correctly: https://github.com/GoogleContainerTools/jib/pull/733/commits/6df4b42d37cc9359a1364fe6eedebc913648808b

I'm fine with either ways.